### PR TITLE
Bump pipelines version 0.12.1

### PR DESCRIPTION
**Pull request by praveen4g0**
Added triggers E2E tests
Updated specs to run downward compatibility tests
Added Dockerfile

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+ENV GOLANG_VERSION 1.14.3
+
+RUN curl -sSL https://storage.googleapis.com/golang/go$GOLANG_VERSION.linux-amd64.tar.gz \
+		| tar -v -C /usr/local -xz
+
+ENV PATH /usr/local/go/bin:$PATH
+RUN mkdir -p /go/src /go/bin && chmod -R 777 /go
+ENV GOROOT /usr/local/go
+ENV GOPATH /go
+ENV PATH /go/bin:$PATH
```
Comment (vdemeester) [inline Dockerfile:24]: We could use `golang:1.14` image here instead :angel:

File: Makefile [hunk 2]
```diff
@@ -1,23 +0,0 @@
-##########------------------------------------------------------------##########
```
Comment (vdemeester) [inline Makefile:1]: Any reason for removing the `Makefile` :stuck_out_tongue_closed_eyes:

File: Dockerfile (hunk 1 above)
Comment (praveen4g0) [reply Dockerfile:24]: yeah, Need to optimise this Dockerfile, I will update that with base image `golang:1.14` and push changes

File: Makefile (hunk 2 above)
Comment (praveen4g0) [reply Makefile:1]: As veeresh is coming up with new model to test cli e2e, `Makefile` is no longer valid here

File: pkg/oc/oc.go [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+import (
+	"log"
+
+	"github.com/openshift-pipelines/release-tests/pkg/cmd"
+	resource "github.com/openshift-pipelines/release-tests/pkg/config"
+)
+
+// Create resources using oc command
+func Create(path_dir, namespace string) {
+	log.Printf("output: %s\n", cmd.MustSucceed("oc", "create", "-f", resource.Path(path_dir), "-n", namespace).Stdout())
```
Comment (VeereshAradhya) [inline pkg/oc/oc.go:12]: better to return the output.

File: specs/pipelines/fail.spec [hunk 4]
```diff
@@ -5,25 +5,38 @@ Contains negative scenarios that exercises running pipeline
...
-Running a task using a ServiceAccount that does not exist must fail
+Running a pipeline using a ServiceAccount that does not exist must fail
 
 Steps:
-  * Create task
   * Verify ServiceAccount "foobar" does not exist
-  * Run task using "foobar" ServiceAccount
-  * Verify taskrun has failed
-
+  * Create
```
Comment (VeereshAradhya) [inline specs/pipelines/fail.spec:15]: `create resource from table` will be descriptive

File: steps/cli/oc.go [hunk 5]
```diff
@@ -0,0 +1,22 @@
+package cli
+
+import (
+	"github.com/getgauge-contrib/gauge-go/gauge"
+	m "github.com/getgauge-contrib/gauge-go/models"
+	"github.com/openshift-pipelines/release-tests/pkg/oc"
+	"github.com/openshift-pipelines/release-tests/pkg/store"
+)
+
+var _ = gauge.Step("Create <table>", func(table *m.Table) {
```
Comment (VeereshAradhya) [inline steps/cli/oc.go:10]: Isn't it better if this take resource location as parameter and the gauge takes care of parameterization? For example if I want to create only one resource for my test case and with the above step implementation we have to write a table for the single resource as well

File: steps/cli/oc.go [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+	for _, row := range table.Rows {
+		resource := row.Cells[1]
+		oc.Create(resource, store.Namespace())
+	}
+})
+
+var _ = gauge.Step("Delete <table>", func(table *m.Table) {
```
Comment (VeereshAradhya) [inline steps/cli/oc.go:17]: same here. The gauge supports parameterization right ?

File: pkg/oc/oc.go (hunk 3 above)
Comment (praveen4g0) [reply pkg/oc/oc.go:12]: As we do all our assertions inside `cmd.Mustsucced`, so I see no value in returning output (atleast for my use case)

File: specs/pipelines/fail.spec (hunk 4 above)
Comment (praveen4g0) [reply specs/pipelines/fail.spec:15]: Agree

File: steps/cli/oc.go (hunk 5 above)
Comment (praveen4g0) [reply steps/cli/oc.go:10]: As per my use case I do perform same action with 2 different resources version `v1alpha1` & `v1beta1`  so I choose data driven approach here to fasten my execution, and also for creating multiple resources, we no need to duplicate steps

File: pkg/oc/oc.go (hunk 3 above)
Comment (praveen4g0) [reply pkg/oc/oc.go:12]: yes `cmd.Mustsucced` will assert for exit code `0` and returns result

File: steps/cli/oc.go (hunk 5 above)
Comment (VeereshAradhya) [reply steps/cli/oc.go:10]: I agree, my point is can we have a step like `create resource "resourceLocation"` and create a data table to create multiple resources? By this way the step can be used to create single resource as well.

Comment (VeereshAradhya) [reply steps/cli/oc.go:10]: Like `Dynamic Parameters` in https://docs.gauge.org/writing-specifications.html?os=linux&language=python&ide=vscode

Comment (praveen4g0) [reply steps/cli/oc.go:10]: yes, gauge has concept `Step alias` is a feature when there are multiple step names for the same step functionality.
eg: https://docs.gauge.org/writing-specifications.html?os=linux&language=java&ide=null

File: Dockerfile (hunk 1 above)
Comment (praveen4g0) [reply Dockerfile]: Updated :+1:

File: Dockerfile [hunk 7]
```diff
@@ -0,0 +1,33 @@
+FROM golang:1.14
+
+
+ENV GO111MODULE on
```
Comment (ppitonak) [inline Dockerfile:4]: is this necessary?

File: Dockerfile [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+    gauge install xml-report && \
+	gauge install reportportal && \
+    gauge config check_updates false && \
+    gauge config runner_connection_timeout 3600000 && \
+    gauge config runner_request_timeout 3600000 && \
+    gauge config ide_request_timeout 3600000 && \
+    gauge config plugin_connection_timeout 3600000 && \
+    gauge config plugin_kill_timeout 3600000
+
+ENV PATH=$HOME/.gauge:$PATH
```
Comment (ppitonak) [inline Dockerfile:28]: What is this good for?

File: Dockerfile [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+
+# Install latest oc and kubectl clients
+ADD scripts/openshift-install.sh /usr/local/bin/
```
Comment (ppitonak) [inline Dockerfile:31]: The name of the script is misleading. What about `oc-install.sh` or `openshift-client-install.sh`

File: env/default/default.properties [hunk 10]
```diff
@@ -10,31 +10,31 @@ gauge_reports_dir = reports
...
-screenshot_on_failure = true
+screenshot_on_failure = false
 
 # The path to the gauge logs directory. Should be either relative to the project directory or an absolute path
 logs_directory = logs
 
 # Timeout in milliseconds for requests from language
 # runner when connecting to IDE.
-ide_request_timeout = 1200000
+ide_request_timeout = 3600000
```
Comment (ppitonak) [inline env/default/default.properties:20]: Why did you change these timeouts?

File: pkg/config/config.go [hunk 11]
```diff
@@ -154,3 +154,19 @@ func DltTempDir() {
...
+	if _, err := os.Stat(td); os.IsNotExist(err) {
+		assert.NoError(err, fmt.Sprintf("Error: in identifying test data path %+v ", td))
+	}
+	return filepath.Join(append([]string{td}, elem...)...)
+}
+
+func ReadBytes(elem string) ([]byte, error) {
+	bytes, err := ioutil.ReadFile(Path(elem))
+	if err != nil {
+		return nil, fmt.Errorf("couldn't load testdata example PullRequest event data: %v", err)
```
Comment (ppitonak) [inline pkg/config/config.go:169]: Please choose one style and use it everywhere (start with upper case, testdata -> test data)

File: specs/triggers/cron.spec [hunk 12]
```diff
@@ -1,20 +1,24 @@
...
-  * Verify resources are created with labels & event-id
-  * Verify pipelinerun is "successfull"
\ No newline at end of file
+  * Create
+    |S.NO|resource_dir                                 |
+    |----|---------------------------------------------|
+    |1   |testdata/triggers/cron/example-pipeline.yaml|
+    |2   |testdata/triggers/cron/tiggerbinding.yaml|
+    |3   |testdata/triggers/cron/triggertemplate.yaml|
+    |4   |testdata/triggers/cron/eventlistener.yaml|
```
Comment (ppitonak) [inline specs/triggers/cron.spec:18]: I like this tabular form, it's much easier to read. Could we have all the last `|` aligned?

**Comment by ppitonak** [comment]
Please update README

File: pkg/config/config.go (hunk 11 above)
Comment (praveen4g0) [reply pkg/config/config.go:169]: Agree

File: env/default/default.properties (hunk 10 above)
Comment (praveen4g0) [reply env/default/default.properties:20]: Sometimes I see due to n/w issues tests fails due to time out so I kept a bigger value, I would change something to reasonable value

File: specs/triggers/cron.spec (hunk 12 above)
Comment (praveen4g0) [reply specs/triggers/cron.spec:18]: issue with my ide will fix indentation

File: Dockerfile (hunk 8 above)
Comment (praveen4g0) [reply Dockerfile]: Ahh, good one this is where all gauge related failure screenshots or other temp files will be created which force me to run as root user I did remove that and tested locally working fine

File: Dockerfile (hunk 9 above)
Comment (praveen4g0) [reply Dockerfile]: updated with `openshift-client-install.sh`

File: specs/pipelines/fail.spec (hunk 4 above)
Comment (praveen4g0) [reply specs/pipelines/fail.spec:15]: I will update it on next pr

**Comment by praveen4g0** [comment]
> Please update README

Updated `README.md` file
